shared/mad/codeql/mad/modelgenerator/internal/ModelGeneratorImpl.qll

```diff
@@ -658,20 +657,40 @@ module MakeModelGenerator<
     ) {
       PropagateContentFlow::flow(p, reads, returnNodeExt, stores, preservesValue) and
       returnNodeExt.getEnclosingCallable() = api and
-      p.(NodeExtended).getEnclosingCallable() = api
+      p.(NodeExtended).getEnclosingCallable() = api and
+      p = api.getARelevantParameterNode()
     }
 
     /**
      * Holds if any of the content sets in `path` translates into a synthetic field.
      */
     private predicate hasSyntheticContent(PropagateContentFlow::AccessPath path) {
-      exists(PropagateContentFlow::AccessPath tail, DataFlow::ContentSet head |
-        head = path.getHead() and
-        tail = path.getTail()
-      |
-        exists(getSyntheticName(head)) or
-        hasSyntheticContent(tail)
-      )
+      exists(getSyntheticName(path.getAtIndex(_)))
+    }
+
+    private string getHashAtIndex(PropagateContentFlow::AccessPath ap, int i) {
+      result = getSyntheticName(ap.getAtIndex(i))
+    }
+
+    private string getReversedHash(PropagateContentFlow::AccessPath ap) {
+      result = concat(int i | | getHashAtIndex(ap, i), "" order by i desc)
+    }
+
+    private string getHash(PropagateContentFlow::AccessPath ap) {
+      result = concat(int i | | getHashAtIndex(ap, i), "" order by i)
+    }
+
+    /**
+     * Gets all access paths that contains the synthetic fields
+     * from `ap` in reverse order (if `ap` contains at least one synthetic field).
+     * These are the possible candidates for synthetic path continuations.
+     */
+    private PropagateContentFlow::AccessPath getSyntheticPathCandidate(
+      PropagateContentFlow::AccessPath ap
+    ) {
+      hasSyntheticContent(ap) and
+      hasSyntheticContent(result) and
```

Comment:
You can use `strictconcat` in `get(Reversed)Hash` an get rid of these two conjuncts.

Reply:
Yes! I also thought about that when writing the code, but I wasn't sure that the code would be as "readable" since the fact ensured by the `hasSyntheticContent` check would be "hidden" in the semantics of `strictconcat`.
Since you find it more readable to use `strictconcat` we should definitely go with that :-)

Reply:
Or is there also a performance reason to use `strictconcat`?